### PR TITLE
feat: Add tracking of run settings

### LINKED ISSUE #4237: Add better tracking of run settings
### Description of the feature

Currently, we do not track specific settings when running a model, which can often influence the results.

My suggestion would be to add the following to the results:

```
# current results
|- model/revision
    |- taskname1.json
    |- ...
    |- model_meta.json
    |- run_settings.jsonl
```

with run_settings.jsonl being the run settings for each task:

```json
{"task": "taskname1", "version": {"mteb": 1.2.3, "torch": 2.8.2, "flash-attn": None, ...}, "encode_kwargs" {...}}
{"task": "taskname2", "version": {"mteb": 1.2.3, "torch": 2.8.2, "flash-attn": None, ...}, "encode_kwargs" {...}}
```

This would be great for debugging. It could be stored in other ways, this is just one suggestion. It could also be stored in the task results which might be easier, but expands the size of an object that we already use a lot.

Let me know what you think

### PR DESCRIPTION
**Pull request by ayush1298**
closes #4237

These PR added a functionality to create `run_settings.jsonl` file to track run settings. 
It will generate `run_settings.jsonl` files per `model/revision` and create one entry per `task/split/subset` combination.

This PR is created with help of Github Copilot.

### CODE REVIEW

File: mteb/evaluate.py [hunk 1]
```diff
@@ -56,6 +57,24 @@ class OverwriteStrategy(HelpfulStrEnum):
     ONLY_CACHE = "only-cache"
 
 
+def _get_package_versions() -> dict[str, str | None]:
+    """Collect version information of key packages used during evaluation.
+
+    Returns:
+        A dictionary with package names as keys and versions (or None if not installed) as values.
+    """
+    packages = ["mteb", "torch", "sentence-transformers", "flash-attn", "transformers"]
```
Comment (ayush1298) [inline mteb/evaluate.py]: I am not sure if we have to use these packages only, or add any other package. Also, should we include package from `extra_requirement_groups` for a specific model?

File: mteb/cache/result_cache.py [hunk 2]
```diff
@@ -346,6 +347,24 @@ def save_to_cache(
...
+                        split=split,
+                        subset=hf_subset,
+                        version=task_result.version or {},
+                        encode_kwargs=task_result.encode_kwargs or {},
+                    )
+                    run_settings_list.append(run_settings)
+
+            if run_settings_list:
+                run_settings_path = result_path.parent / "run_settings.jsonl"
+                append_run_settings_to_file(run_settings_path, run_settings_list)
```
[the remark this answers is not in the document]
Comment (ayush1298) [reply mteb/cache/result_cache.py]: fix

File: mteb/evaluate.py (hunk 1 above)
Comment (Samoed) [reply mteb/evaluate.py]: I think we can start from these packages. I don't think that this is nessarty to include from `extra_requirement_groups`

File: mteb/results/task_result.py [hunk 3]
```diff
@@ -1020,3 +1035,89 @@ class TaskError(BaseModel):
...
+        """Create RunSettings from a JSON line (JSONL format).
+
+        Args:
+            line: A JSON string from JSONL format.
+
+        Returns:
+            The created RunSettings object.
+        """
+        data = json.loads(line)
+        return cls.model_validate(data)
```
Comment (Samoed) [inline mteb/results/task_result.py]:
```suggestion
        return cls.model_validate_json(line)
```

File: mteb/results/task_result.py [hunk 4]
```diff
@@ -1020,3 +1035,89 @@ class TaskError(BaseModel):
...
+    version: dict[str, str | None]
+    encode_kwargs: dict[str, Any]
+
+    def to_json_line(self) -> str:
+        """Convert RunSettings to a JSON line (JSONL format).
+
+        Returns:
+            A JSON string suitable for JSONL format (single line).
+        """
+        return json.dumps(self.model_dump(), separators=(",", ": "))
```
Comment (Samoed) [inline mteb/results/task_result.py]:
```suggestion
        return self.model_dump_json()
```

File: mteb/results/task_result.py [hunk 5]
```diff
@@ -172,6 +175,8 @@ class TaskResult(BaseModel):  # noqa: PLR0904
     evaluation_time: float | None
     kg_co2_emissions: float | None = None
     date: datetime.datetime | None = None
+    version: dict[str, str | None] | None = None
+    encode_kwargs: dict[str, Any] | None = None
```
Comment (Samoed) [inline mteb/results/task_result.py]: I don't think that we should save info of libraries in `version` and that we should save `encode_kwargs` in task results

File: mteb/evaluate.py (hunk 1 above)
Comment (ayush1298) [reply mteb/evaluate.py]: cool !

File: mteb/results/task_result.py (hunk 5 above)
Comment (ayush1298) [reply mteb/results/task_result.py]: How should we save them then?

Comment (Samoed) [reply mteb/results/task_result.py]: We can get them directly from env before dump jsonl

Comment (ayush1298) [reply mteb/results/task_result.py]: We can get libraries version from environment, how we can get `encode_kwargs?`
`evaluate` has `encode_kwargs`, can we pass it to `cache.save_to_cache` and from there we can pass to `RunSettings?`

Comment (Samoed) [reply mteb/results/task_result.py]: I think yes, we can both libraries and encode kwargs to there

Comment (ayush1298) [reply mteb/results/task_result.py]: I have fixed it. Can you check?

File: mteb/results/task_result.py [hunk 6]
```diff
@@ -1020,3 +1021,88 @@ class TaskError(BaseModel):
 
     task_name: str
     exception: str
+
+
+class RunSettings(BaseModel):
```
Comment (KennethEnevoldsen) [inline mteb/results/task_result.py]: I would move it to a seperate module

File: mteb/results/task_result.py [hunk 7]
```diff
@@ -1020,3 +1021,88 @@ class TaskError(BaseModel):
...
+        Args:
+            line: A JSON string from JSONL format.
+
+        Returns:
+            The created RunSettings object.
+        """
+        return cls.model_validate_json(line)
+
+
+def append_run_settings_to_file(
```
Comment (KennethEnevoldsen) [inline mteb/results/task_result.py]: what if I rerun a split?

File: mteb/evaluate.py [hunk 8]
```diff
@@ -366,7 +366,11 @@ def evaluate(  # noqa: PLR0913, PLR0914
         )
         combined_results = aggregated_task.combine_task_results(results.task_results)
         if cache:
-            cache.save_to_cache(combined_results, meta)
+            cache.save_to_cache(
+                combined_results,
+                meta,
+                encode_kwargs=cast("dict[str, Any] | None", encode_kwargs),
```
Comment (KennethEnevoldsen) [inline mteb/evaluate.py]: cast should not be in the call

Comment (KennethEnevoldsen) [reply mteb/evaluate.py]: same below (typing should be on their own line)

File: mteb/cache/result_cache.py [hunk 9]
```diff
@@ -346,6 +352,61 @@ def save_to_cache(
             with model_meta_path.open("w") as f:
                 json.dump(meta.to_dict(), f, default=str, indent=4)
 
+        version_dict = self._get_package_versions()
+
+        run_settings_list = []
+        for split, split_scores in task_result.scores.items():
+            for score_entry in split_scores:
+                hf_subset = score_entry.get("hf_subset", "default")
+                run_settings = RunSettings(
```
Comment (KennethEnevoldsen) [inline mteb/cache/result_cache.py]: I am not sure we need an object here. I think we just need a `json_serialize_kwargs` function and `write_to_keyed_json` function. If we run the same model twice we don't want duplicate entries - we need a test for this.

Note that encode_kwargs is not always json serializable (we probably need a test to ensure that we can handle these)

File: mteb/cache/result_cache.py [hunk 10]
```diff
@@ -346,6 +352,61 @@ def save_to_cache(
...
+                for rs in existing_entries
+                if (rs.task, rs.split, rs.subset) not in new_keys
+            ]
+            all_entries = filtered_existing + run_settings_list
+            with run_settings_path.open("w", encoding="utf-8") as f:
+                for entry in all_entries:
+                    f.write(entry.to_json_line() + "\n")
+
+    @staticmethod
+    def _get_package_versions() -> dict[str, str | None]:
```
Comment (KennethEnevoldsen) [inline mteb/cache/result_cache.py]: this should not be on the object just place it above.

File: mteb/cache/result_cache.py [hunk 11]
```diff
@@ -346,6 +352,61 @@ def save_to_cache(
...
+        """Get current package versions from the environment.
+
+        Returns:
+            A dictionary with package names as keys and versions (or None if not installed) as values.
+        """
+        from importlib.metadata import version
```
Comment (KennethEnevoldsen) [inline mteb/cache/result_cache.py]: top-level (I feel like I keep mentioning this - packages should always be top level if possible, especially for native modules)

**Review by KennethEnevoldsen — CHANGES_REQUESTED** [review]

File: mteb/cache/result_cache.py (hunk 10 above)
Comment (ayush1298) [reply mteb/cache/result_cache.py]: done

File: mteb/results/task_result.py (hunk 6 above)
Comment (ayush1298) [reply mteb/results/task_result.py]: removing class

File: mteb/evaluate.py (hunk 8 above)
Comment (ayush1298) [reply mteb/evaluate.py]: done

File: mteb/results/task_result.py (hunk 7 above)
Comment (ayush1298) [reply mteb/results/task_result.py]: this is already handled. If you see `write_to_keyed_json` function, it will check for new and existing entries and merged them correctly

File: mteb/cache/result_cache.py (hunk 9 above)
Comment (ayush1298) [reply mteb/cache/result_cache.py]: agree. I have removed object and add helper functions.
Also added test for duplicate entries and serialization of `encode_kwargs`

File: mteb/results/task_result.py [hunk 12]
```diff
@@ -1020,3 +1021,52 @@ class TaskError(BaseModel):
 
     task_name: str
     exception: str
+
+
+def json_serialize_kwargs(kwargs: Any | None) -> Any:
+    """Convert keyword arguments into a JSON-serializable structure."""
+    return json.loads(json.dumps(kwargs, default=str)) if kwargs is not None else {}
```
Comment (Samoed) [inline mteb/results/task_result.py]: But wouldn't `json.dump` will do the same later? Also, there can be an error, so we should add try except on this

File: mteb/results/task_result.py [hunk 13]
```diff
@@ -1020,3 +1021,52 @@ class TaskError(BaseModel):
 
     task_name: str
     exception: str
+
+
+def json_serialize_kwargs(kwargs: Any | None) -> Any:
```
Comment (Samoed) [inline mteb/results/task_result.py]:
```suggestion
def json_serialize_kwargs(kwargs: dict[str, Any] | None) -> dict[str, Any]:
```

File: mteb/results/task_result.py [hunk 14]
```diff
@@ -1020,3 +1021,52 @@ class TaskError(BaseModel):
...
+            if isinstance(parsed, dict):
+                run_settings.append(parsed)
+    return run_settings
+
+
+def write_to_keyed_json(
+    path: Path,
+    entries: list[dict[str, Any]],
+    *,
+    key_fields: tuple[str, ...] = ("task", "split", "subset"),
```
Comment (Samoed) [inline mteb/results/task_result.py]: Is it expecting to be changed?

**Review by Samoed — COMMENTED** [review]
Can you make your new helping function private?

File: mteb/results/task_result.py (hunk 14 above)
Comment (ayush1298) [reply mteb/results/task_result.py]: no, I guess.

File: mteb/results/task_result.py (hunk 12 above)
Comment (ayush1298) [reply mteb/results/task_result.py]: added try-except. 
I guess we have to serialize `encode_kwargs` before we can use it in `json.dump`

File: mteb/results/task_result.py (hunk 13 above)
Comment (ayush1298) [reply mteb/results/task_result.py]: using `kwargs: Mapping[str, Any] | None` to match typing

File: mteb/results/task_result.py [hunk 15]
```diff
@@ -1020,3 +1021,59 @@ class TaskError(BaseModel):
 
     task_name: str
     exception: str
+
+
+def _json_serialize_kwargs(kwargs: Mapping[str, Any] | None) -> dict[str, Any]:
```
Comment (KennethEnevoldsen) [inline mteb/results/task_result.py]: @samoed don't we already have function for this in the experiments - seems like one that we should re-use?

File: tests/test_run_settings_jsonl.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def test_save_to_cache_replaces_existing_run_settings_entry(tmp_path, monkeypatch):
+    cache = ResultCache(cache_path=tmp_path)
+    task_result = DummyTaskResult(
+        task_name="STS12",
+        scores={"test": [{"hf_subset": "en"}]},
+    )
+
+    monkeypatch.setattr(
```
Comment (KennethEnevoldsen) [inline tests/test_run_settings_jsonl.py]: why the monkeypatch here?

File: tests/test_run_settings_jsonl.py [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def _read_jsonl(path):
+    with path.open("r", encoding="utf-8") as handle:
+        return [json.loads(line) for line in handle if line.strip()]
+
+
+def test_save_to_cache_replaces_existing_run_settings_entry(tmp_path, monkeypatch):
+    cache = ResultCache(cache_path=tmp_path)
+    task_result = DummyTaskResult(
```
Comment (KennethEnevoldsen) [inline tests/test_run_settings_jsonl.py]: Why not use a actual task result object?

File: mteb/results/task_result.py [hunk 18]
```diff
@@ -1020,3 +1021,59 @@ class TaskError(BaseModel):
...
+    """Convert keyword arguments into a JSON-serializable structure."""
+    if kwargs is None:
+        return {}
+
+    try:
+        return json.loads(json.dumps(kwargs, default=str))
+    except Exception as e:
+        logger.warning(f"Could not serialize encode_kwargs to JSON: {e}")
+        return {}
+
```
Comment (KennethEnevoldsen) [inline mteb/results/task_result.py]: No need for the try except - it should never fail
```
def make_json_serializable(kwargs: dict | None) -> dict:
    if kwargs is None:
        return {}
    return json.loads(json.dumps(kwargs, default=str))
```

File: tests/test_run_settings_jsonl.py [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+        self.scores = scores
+
+    @staticmethod
+    def to_disk(path) -> None:
+        path.write_text("{}", encoding="utf-8")
+
+
+class UnserializableValue:
+    def __str__(self) -> str:
+        return "unserializable-value"
```
Comment (KennethEnevoldsen) [inline tests/test_run_settings_jsonl.py]: you don't need the string conversion here - in the test simlpy check for that it is a string (it could seem like here that it is needed but itis not

File: mteb/results/task_result.py (hunk 12 above)
Comment (KennethEnevoldsen) [reply mteb/results/task_result.py]: I don't think there can be an error here (it would be in the string serialization)

File: mteb/results/task_result.py (hunk 18 above)
Comment (KennethEnevoldsen) [reply mteb/results/task_result.py]: (also see comment above)

**Review by KennethEnevoldsen — APPROVED** [review]
we are getting closer. 

Renamed from fix to feat as it adds a feature that user might rely on.

**Review by KennethEnevoldsen — COMMENTED** [review]
(accidentally accepted

File: mteb/results/task_result.py (hunk 15 above)
Comment (ayush1298) [reply mteb/results/task_result.py]: are you referring to this function? 
https://github.com/embeddings-benchmark/mteb/blob/892225eb2981060a8e68cca84c6e28e38631065e/mteb/models/model_meta.py#L1591

File: tests/test_run_settings_jsonl.py (hunk 16 above)
Comment (ayush1298) [reply tests/test_run_settings_jsonl.py]: removed

File: tests/test_run_settings_jsonl.py (hunk 17 above)
Comment (ayush1298) [reply tests/test_run_settings_jsonl.py]: done

**Comment by ayush1298** [comment]
@KennethEnevoldsen I have updated test, and also using `task_result` object from `TaskResult.from_task_results`. Though I am not sure, are you referring to these thing only.

File: mteb/results/task_result.py (hunk 15 above)
Comment (ayush1298) [reply mteb/results/task_result.py]: Its a different one and I think can't be used directly and apart from that I didn't find any other function

Comment (KennethEnevoldsen) [reply mteb/results/task_result.py]: Yeah no it doesn't seem like it - will close this

File: mteb/results/task_result.py [hunk 20]
```diff
@@ -5,7 +5,8 @@
 import logging
 from collections import defaultdict
 from functools import cached_property
-from importlib.metadata import version
+from importlib.metadata import version as get_version
```
Comment (KennethEnevoldsen) [inline mteb/results/task_result.py]: why the rename here?

File: mteb/results/task_result.py [hunk 21]
```diff
@@ -1020,3 +1021,54 @@ class TaskError(BaseModel):
...
+                logger.warning(
+                    f"Could not parse run_settings line '{stripped_line}': {e}"
+                )
+                continue
+            if isinstance(parsed, dict):
+                run_settings.append(parsed)
+    return run_settings
+
+
+def _write_to_keyed_json(
```
Comment (KennethEnevoldsen) [inline mteb/results/task_result.py]:
```suggestion
def _write_and_merge_keyed_json(
```

File: mteb/results/task_result.py [hunk 22]
```diff
@@ -1020,3 +1021,54 @@ class TaskError(BaseModel):
...
+    path: Path,
+    entries: list[dict[str, Any]],
+    *,
+    key_fields: tuple[str, str, str] = ("task", "split", "subset"),
+) -> None:
+    """Write JSONL entries, replacing any existing entries with the same key."""
```
Comment (KennethEnevoldsen) [inline mteb/results/task_result.py]:
```suggestion
    """Write entries to `.jsonl`, if it already exist it will merge it, replacing any existing entries with the same key."""
```

**Review by KennethEnevoldsen — COMMENTED** [review]
I few minor things - I think we need the what's new documentation, but otherwise I think we are good

File: mteb/cache/result_cache.py [hunk 23]
```diff
@@ -50,6 +52,30 @@
 _EXPERIMENTS_FOLDER_NAME = "experiments"
 
 
+def get_package_versions() -> dict[str, str | None]:
```
Comment (Samoed) [inline mteb/cache/result_cache.py]:
```suggestion
def _get_package_versions() -> dict[str, str | None]:
```

File: mteb/cache/result_cache.py [hunk 24]
```diff
@@ -346,6 +374,25 @@ def save_to_cache(
...
+                    "split": split,
+                    "subset": hf_subset,
+                    "version": version_dict,
+                    "encode_kwargs": _json_serialize_kwargs(encode_kwargs),
+                }
+                run_settings_list.append(run_settings)
+
+        if run_settings_list:
+            run_settings_path = result_path.parent / "run_settings.jsonl"
+            _write_to_keyed_json(run_settings_path, run_settings_list)
```
Comment (Samoed) [inline mteb/cache/result_cache.py]: Maybe better to move to other function?

File: mteb/results/task_result.py (hunk 20 above)
Comment (ayush1298) [reply mteb/results/task_result.py]: was required in old implemetation. forgot to revert. Will revert it.

File: mteb/cache/result_cache.py [hunk 25]
```diff
@@ -321,6 +347,7 @@ def save_to_cache(
         task_result: TaskResult,
         model_name: str | ModelMeta,
         model_revision: str | None = None,
+        encode_kwargs: Mapping[str, Any] | None = None,
```
Comment (Samoed) [inline mteb/cache/result_cache.py:373]:
```suggestion
        *,
        encode_kwargs: Mapping[str, Any] | None = None,
```

File: mteb/cache/result_cache.py (hunk 24 above)
Comment (ayush1298) [reply mteb/cache/result_cache.py]: Its already a very small function, do we have to make separate function

**Comment by Samoed** [comment]
I think this `run_expriments.json` wouldn't be handled by `results_cache.submit_results`

**Comment by ayush1298** [comment]
> I think this `run_expriments.json` wouldn't be handled by `results_cache.submit_results`

Changed to handle in `submit_results`

File: mteb/results/task_result.py [hunk 26]
```diff
@@ -1020,3 +1021,54 @@ class TaskError(BaseModel):
 
     task_name: str
     exception: str
+
+
+def _json_serialize_kwargs(kwargs: Mapping[str, Any] | None) -> dict[str, Any]:
+    """Convert keyword arguments into a JSON-serializable structure."""
+    if kwargs is None:
+        return {}
+    return cast("dict[str, Any]", json.loads(json.dumps(kwargs, default=str)))
```
Comment (ayush1298) [inline mteb/results/task_result.py]: typecheck is giving error here, so need to add cast again

Comment (KennethEnevoldsen) [reply mteb/results/task_result.py]: I would generally recommend not having cast and functionality in the same line:


```suggestion
    json_kwargs = json.loads(json.dumps(kwargs, default=str))
    json_kwargs = cast("dict[str, Any]", json_kwargs)
    return json_kwargs
```

File: mteb/results/task_result.py [hunk 27]
```diff
@@ -1020,3 +1021,54 @@ class TaskError(BaseModel):
 
     task_name: str
     exception: str
+
+
+def _json_serialize_kwargs(kwargs: Mapping[str, Any] | None) -> dict[str, Any]:
```
Comment (Samoed) [inline mteb/results/task_result.py]: Do we need this function? We'll use `json.dump` in `_write_and_merge_keyed_json`

Comment (ayush1298) [reply mteb/results/task_result.py]: removed function, but instead of adding to `_write_and_merge_keyed_json,` used it before only

Comment (Samoed) [reply mteb/results/task_result.py]: I meant if we doing `json.dump` later, why do we `json.dump` encode kwargs?

Comment (ayush1298) [reply mteb/results/task_result.py]: it may contain objects that can't be serialized, so I guess we need it

Comment (Samoed) [reply mteb/results/task_result.py]: How this would help? You'll use same json dump, but later

Comment (ayush1298) [reply mteb/results/task_result.py]: inside `_write_and_merge_keyed_json` function, we are not unfolding these `run_settings_list` which contain each entry as a dict inside which we have `encode_kwargs,` so why do all that unfolding stuff again? We can just do it when creating these `run_settings_list`

File: mteb/results/task_result.py (hunk 26 above)
Comment (ayush1298) [reply mteb/results/task_result.py]: I have removed the function. (below comment)

Comment (KennethEnevoldsen) [reply mteb/results/task_result.py]: since we have talked about this a few times I made a rule request on ruff: 
https://github.com/astral-sh/ruff/issues/25215

File: mteb/results/task_result.py (hunk 27 above)
Comment (ayush1298) [reply mteb/results/task_result.py]: @KennethEnevoldsen What's your opinion here?

Comment (Samoed) [reply mteb/results/task_result.py]: > it may contain objects that can't be serialized, so I guess we need it

This wouldn't help, because json.dump, will just raise an error

```py
import torch
import json
json.dumps({"dtype": torch.bfloat16})
# TypeError: Object of type dtype is not JSON serializable when serializing dict item 'dtype'
```

Comment (ayush1298) [reply mteb/results/task_result.py]:
```py
import torch
import json
encode_kwargs = {"dtype": torch.bfloat16}
print(json.dumps(encode_kwargs, default=str))
```

we are using `default=str,` and above code is running correctly

**Comment by ayush1298** [comment]
Is there anything else left here?
I have added its support in `submit_results` and also updated `whats_new` docs

**Review by Samoed — APPROVED** [review]
Looks good to me

**Comment by ayush1298** [comment]
@KennethEnevoldsen Do you want to review once again?

**Review by KennethEnevoldsen — APPROVED** [review]

**Comment by KennethEnevoldsen** [comment]
Good catch on submit_results. I think the PR is looking good!
